File: app/src/main/java/android/mahendra/attendancemanager/TimeTableActivity.java
package android.mahendra.attendancemanager;

import android.content.Context;
import android.content.Intent;
import android.mahendra.attendancemanager.dialogs.PeriodDialogFragment;
import android.mahendra.attendancemanager.models.Period;
import android.mahendra.attendancemanager.viewmodels.PeriodListViewModel;
import android.mahendra.attendancemanager.viewmodels.SubjectViewModel;
import android.os.Bundle;
import android.util.Log;
import android.util.SparseArray;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.ViewModelProvider;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class TimeTableActivity extends AppCompatActivity implements
        DayScheduleFragment.Callbacks, PeriodDialogFragment.Callbacks {
    private static final String TAG = "TimeTableActivity";

    public static SparseArray<String> WEEK_DAYS;

    private ViewPager2 mViewPager;

    private SubjectViewModel mSubjectViewModel;
    private PeriodListViewModel mPeriodListViewModel;

    private List<String> mSubjectsTitles;

    private int mWeekDay = -1;
    private int mWeekDayOffSet = -1;

    private Period mTempPeriod;
    private boolean addNewPeriod;

    public static Intent newIntent(Context context) {
        Intent i = new Intent(context, TimeTableActivity.class);
        return i;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_timetable);
        createWeekDayHash();
        mWeekDay = Calendar.getInstance().get(Calendar.DAY_OF_WEEK);

        mSubjectViewModel = new ViewModelProvider(this).get(SubjectViewModel.class);

        mSubjectViewModel.getSubjectTitles().observe(this, subjectTitles ->
                mSubjectsTitles = subjectTitles);

        mPeriodListViewModel = new ViewModelProvider(this).get(PeriodListViewModel.class);
        mViewPager = findViewById(R.id.day_schedule_viewpager);
        mViewPager.setAdapter(new DayScheduleAdapter(this));
        mViewPager.setCurrentItem(mWeekDay - mWeekDayOffSet);
        TabLayout tabLayout = findViewById(R.id.tab_layout_weekday);
        new TabLayoutMediator(tabLayout, mViewPager,
                (tab, position) -> tab.setText(WEEK_DAYS.get(position + mWeekDayOffSet))
         ).attach();
    }

    private class DayScheduleAdapter extends FragmentStateAdapter {
        List<Integer> weekDays = new ArrayList<>();

        public DayScheduleAdapter(FragmentActivity activity) {
            super(activity);
            for (int i = 1; i < 7; i++) {
                weekDays.add(i + 1);
            }
            mWeekDayOffSet = weekDays.get(0);
        }

        @NonNull
        @Override
        public Fragment createFragment(int position) {
            Log.i(TAG, "createFragment: " + position);
            return DayScheduleFragment.newInstance(weekDays.get(position));
        }

        @Override
        public int getItemCount() {
            return weekDays.size();
        }
    }

    public void openAddPeriodDialog(String periodTitle) {
        PeriodDialogFragment dialogFragment
                = PeriodDialogFragment.newInstance(periodTitle);
        FragmentManager fm = getSupportFragmentManager();
        dialogFragment.show(fm, "period");
    }

    @Override
    public ArrayList<String> getSubjectTitles() {
        return (ArrayList<String>) mSubjectsTitles;
    }

    @Override
    public void onAddPeriod(int periodNumber, int weekDay) {
        addNewPeriod = true;
        mTempPeriod = new Period(null, periodNumber, weekDay);
        openAddPeriodDialog(mTempPeriod.getSubjectTitle());
    }

    @Override
    public void onModifyPeriod(Period period) {
        Log.i(TAG, "onModifyPeriod: " + period.getSubjectTitle());
        addNewPeriod = false;
        mTempPeriod = period;
        openAddPeriodDialog(period.getSubjectTitle());
    }

    @Override
    public void onDeletePeriod(String title) {
        Log.i(TAG, "onDeletePeriod: " + mTempPeriod);
        mPeriodListViewModel.deletePeriod(mTempPeriod.getPeriodNumber(), mTempPeriod.getWeekDay());
    }

    @Override
    public void onPeriodSelected(String title) {
        mTempPeriod.setSubjectTitle(title);
        if (addNewPeriod) {
            mPeriodListViewModel.insert(mTempPeriod);
        }
        else {
            mPeriodListViewModel.update(mTempPeriod);
        }
        mTempPeriod = null;
        addNewPeriod = false;
    }

    private void createWeekDayHash() {
        WEEK_DAYS = new SparseArray<>();
        WEEK_DAYS.put(1, "sunday");
        WEEK_DAYS.put(2, "monday");
        WEEK_DAYS.put(3, "tuesday");
        WEEK_DAYS.put(4, "wednesday");
        WEEK_DAYS.put(5, "thursday");
        WEEK_DAYS.put(6, "friday");
        WEEK_DAYS.put(7, "saturday");
    }
}

File: app/src/main/java/android/mahendra/attendancemanager/TimeTableActivity.kt
package android.mahendra.attendancemanager

import android.content.Context
import android.content.Intent
import android.mahendra.attendancemanager.dialogs.PeriodDialogFragment
import android.mahendra.attendancemanager.fragments.DayScheduleFragment
import android.mahendra.attendancemanager.models.Period
import android.mahendra.attendancemanager.viewmodels.PeriodListViewModel
import android.mahendra.attendancemanager.viewmodels.SubjectViewModel
import android.os.Bundle
import android.util.SparseArray
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2

import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import kotlinx.android.synthetic.main.activity_timetable.*

import java.util.Calendar
import kotlin.collections.ArrayList
import kotlin.collections.List

class TimeTableActivity : AppCompatActivity(),
        DayScheduleFragment.Callbacks, PeriodDialogFragment.Callbacks {
    private lateinit var mViewPager: ViewPager2

    private lateinit var mSubjectViewModel: SubjectViewModel
    private lateinit var mPeriodListViewModel: PeriodListViewModel

    private lateinit var mSubjectsTitles: List<String>

    override val subjectTitles: List<String>
        get() = mSubjectsTitles

    private var mWeekDay = -1
    private var mWeekDayOffSet = -1

    private var mTempPeriod: Period? = null
    private var addNewPeriod: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_timetable)
        createWeekDayHash()
        mWeekDay = Calendar.getInstance().get(Calendar.DAY_OF_WEEK)

        mSubjectViewModel = ViewModelProvider(this).get(SubjectViewModel::class.java)

        mSubjectViewModel.getSubjectTitles().observe(this,
                Observer { subjectTitles: List<String> -> mSubjectsTitles = subjectTitles })

        mPeriodListViewModel = ViewModelProvider(this).get(PeriodListViewModel::class.java)
        mViewPager = day_schedule_viewpager
        mViewPager.adapter = DayScheduleAdapter(this)
        mViewPager.currentItem = mWeekDay - mWeekDayOffSet
        val tabLayout: TabLayout = tab_layout_weekday
        TabLayoutMediator(tabLayout, mViewPager
        ) { tab, position -> tab.text = WEEK_DAYS.get(position + mWeekDayOffSet)}.attach()
    }

    private inner class DayScheduleAdapter(fragmentActivity: FragmentActivity)
        : FragmentStateAdapter(fragmentActivity) {

        var weekDays: MutableList<Int> = ArrayList()
        init {
            for (i in 2..7) {
                weekDays.add(i)
            }
            mWeekDayOffSet = weekDays[0]
        }

        override fun createFragment(position: Int): Fragment {
            return DayScheduleFragment.newInstance(weekDays[position])
        }

        override fun getItemCount(): Int {
            return weekDays.size
        }
    }

    private fun openAddPeriodDialog(periodTitle: String?) {
        val dialogFragment = PeriodDialogFragment.newInstance(periodTitle)
        val fm = supportFragmentManager
        dialogFragment.show(fm, "period")
    }

    override fun onAddPeriod(periodNumber: Int, weekDay: Int) {
        addNewPeriod = true
        mTempPeriod = Period(null, periodNumber, weekDay)
        openAddPeriodDialog(mTempPeriod!!.subjectTitle)
    }

    override fun onModifyPeriod(period: Period) {
        addNewPeriod = false
        mTempPeriod = period
        openAddPeriodDialog(period.subjectTitle)
    }

    override fun onDeletePeriod(title: String?) {
        mPeriodListViewModel.deletePeriod(mTempPeriod!!.periodNumber, mTempPeriod!!.weekDay)
    }

    override fun onPeriodSelected(title: String?) {
        mTempPeriod!!.subjectTitle = title
        if (addNewPeriod) {
            mPeriodListViewModel.insert(mTempPeriod!!)
        }
        else {
            mPeriodListViewModel.update(mTempPeriod!!)
        }
        mTempPeriod = null
        addNewPeriod = false
    }

    companion object {
        private const val TAG = "TimeTableActivity"
        lateinit var WEEK_DAYS: SparseArray<String>

        fun newIntent(context: Context): Intent {
            return Intent(context, TimeTableActivity::class.java)
        }

        private fun createWeekDayHash() {
            WEEK_DAYS = SparseArray()
            WEEK_DAYS.put(1, "sunday")
            WEEK_DAYS.put(2, "monday")
            WEEK_DAYS.put(3, "tuesday")
            WEEK_DAYS.put(4, "wednesday")
            WEEK_DAYS.put(5, "thursday")
            WEEK_DAYS.put(6, "friday")
            WEEK_DAYS.put(7, "saturday")
        }
    }
}
